content/posts/i-like-rust.md

```diff
@@ -0,0 +1,218 @@
+But the effective cost difference is huge.
+IBM showed this in [a study](https://www.researchgate.net/publication/255965523_Integrating_Software_Assurance_into_the_Software_Development_Life_Cycle_SDLC) conducted in 2010.
+
+![IBM System Science Institute. A graph depicting the relative cost of fixing defects. "Design" has a cost of 1, "implementation" a cost of 6.5, "testing" a cost of 15 and "maintenance" a cost of 100.](/ibm-relative-cost-of-fixing-defects.png)
+
+Rust is as statically typed and type safe as you could possibly get with a "real" programming language. Static types don't automatically lead to type safety. For example in Java you can do the following to trick the compiler and cause a runtime error:
+
+```java
+int x = 42;
+Object forgetCompileTypeInfo = x; // everything is an object
+return (String) forgetCompileTypeInfo; // oh no, a runtime error
+```
+
+In Rust that's not possible. Casts must be valid at the type level, or else they will be prevented statically.[^3] Not checking for type constraints at compile time, or worse, doing conversions at runtime to somehow satisfy type constraints, as JavaScript does[^4], leads to more bugs and costs.
+
+## No billion dollar bug
+
+Many languages have incorporated the concept of lack of value into the language, with special keywords like `null`, `nil` or similar. Any time you pass a value by reference, which is the default in many languages (Java, C#, Python, ...), the value can be absent. This concept was introduced by Tony Hoare into ALGOL and has become the norm for most programming languages. However, introducing such a special value that can be present everywhere in the system by default, means you have to handle the special value everywhere with additional code, otherwise you might create bugs. Conventions and best practices started to appear, about when to use this special value. Over time people started to agree that this special value shouldn't be everywhere by default. The inventor himself apologised for introducing the concept in the first place.
+
+> But I couldn't resist the temptation to put in a null reference, simply because it was so easy to implement. This has led to innumerable errors, vulnerabilities, and system crashes, which have probably caused a billion dollars of pain and damage in the last forty years.[^5]
+
+Rust is one of the languages which does not know a special null value. You must opt in for the possible absence of a value. This is done with the `Option<T>` type.
+
+## Error handling
+
+Most programmers today agree that using the GOTO statement in high-level languages is bad practice. It leads code that is more difficult to read and maintain. I think exceptions like they are known in many modern languages are just fancy GOTO statements. When an exception is hit, the normal execution flow is stopped and the instruction pointer jumps into some error handling section.
```

Comment:
> Most programmers today agree that using the GOTO statement in high-level languages is bad practice.

Maybe link to the paper? https://homepages.cwi.nl/~storm/teaching/reader/Dijkstra68.pdf